cassandra/connection.py

```diff
@@ -599,6 +600,39 @@ def int_from_buf_item(i):
 else:
     int_from_buf_item = ord
 
+class ShardingInfo(object):
+
+    def __init__(self, shard_id, shards_count, partitioner, sharding_algorithm, sharding_ignore_msb):
+        self.shards_count = int(shards_count)
+        self.partitioner = partitioner
+        self.sharding_algorithm = sharding_algorithm
+        self.sharding_ignore_msb = int(sharding_ignore_msb)
+
+    @staticmethod
+    def parse_sharding_info(message):
+        shard_id = message.options.get('SCYLLA_SHARD', [''])[0] or None
+        shards_count = message.options.get('SCYLLA_NR_SHARDS', [''])[0] or None
+        partitioner = message.options.get('SCYLLA_PARTITIONER', [''])[0] or None
+        sharding_algorithm = message.options.get('SCYLLA_SHARDING_ALGORITHM', [''])[0] or None
+        sharding_ignore_msb = message.options.get('SCYLLA_SHARDING_IGNORE_MSB', [''])[0] or None
+
+        if not (shard_id or shards_count or partitioner == "org.apache.cassandra.dht.Murmur3Partitioner" or
+            sharding_algorithm ==  "biased-token-round-robin" or sharding_ignore_msb):
+            return 0, None
+
+        return int(shard_id), ShardingInfo(shard_id, shards_count, partitioner, sharding_algorithm, sharding_ignore_msb)
+
+    def shard_id(self, t):
+        token = t.value
+        token += MIN_LONG
+        token <<= self.sharding_ignore_msb
+        tokLo = token & 0xffffffff
+        tokHi = (token >> 32) & 0xffffffff
+        mul1 = tokLo * self.shards_count
+        mul2 = tokHi * self.shards_count
+        _sum = (mul1 >> 32) + mul2
+        output = _sum >> 32
+        return output
```

Comment:
Python integers are arbitrary in range, I think that the equivalent of 128-bit C code can be used instead.

Reply:
Ugh, I just notice that this PR is closed and #6 is the follow-up, reposting there.